Tools/scripts/stable_abi.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+
+import argparse
+import glob
+import re
+import pathlib
+import subprocess
+import sys
+import sysconfig
+
+EXCLUDED_HEADERS = {
+    "bytes_methods.h",
+    "cellobject.h",
+    "classobject.h",
+    "code.h",
+    "compile.h",
+    "datetime.h",
+    "dtoa.h",
+    "frameobject.h",
+    "funcobject.h",
+    "genobject.h",
+    "longintrepr.h",
+    "parsetok.h",
+    "pyarena.h",
+    "pyatomic.h",
+    "pyctype.h",
+    "pydebug.h",
+    "pytime.h",
+    "symtable.h",
+    "token.h",
+    "ucnhash.h",
+}
+
+
+def get_exported_symbols(library, dynamic=False):
+    # Only look at dynamic symbols
+    args = ["nm", "--no-sort"]
+    if dynamic:
+        args.append("--dynamic")
+    args.append(library)
+    proc = subprocess.run(args, stdout=subprocess.PIPE, universal_newlines=True)
+    if proc.returncode:
+        sys.stdout.write(proc.stdout)
+        sys.exit(proc.returncode)
+
+    stdout = proc.stdout.rstrip()
+    if not stdout:
+        raise Exception("command output is empty")
+
+    for line in stdout.splitlines():
+        # Split line '0000000000001b80 D PyTextIOWrapper_Type'
+        if not line:
+            continue
+
+        parts = line.split(maxsplit=2)
+        if len(parts) < 3:
+            continue
+
+        symbol = parts[-1]
+        yield symbol
+
+
+def check_library(library, abi_funcs, dynamic=False):
+    available_symbols = set(get_exported_symbols(library, dynamic))
+    missing_symbols = abi_funcs - available_symbols
+    if missing_symbols:
+        print(
+            f"Some symbols from the stable ABI are missing: {', '.join(missing_symbols)}"
+        )
+        return 1
+    return 0
+
+
+def generate_limited_api_symbols(args):
+    library = sysconfig.get_config_var("LIBRARY")
+    ldlibrary = sysconfig.get_config_var("LDLIBRARY")
+    if ldlibrary != library:
+        raise Exception("Limited ABI symbols can only be generated from a static build")
```

Comment:
What is the issue with the libpython dynamic library?

Recently, I modified Tools/scripts/smelly.py to check not only the static library, but also the libpython dynamic library and C extensions.

Reply:
> What is the issue with the libpython dynamic library?

That the dynamic library may have some symbols missing if is built-in weird ways. Honestly, the reason I am raising here was to keep the script as simple as possible on the first version :)

Reply:
Ok.